File: packages/core/src/params.ts
/**
 * WOTS Parameter Set - BouncyCastle Compatible (w=8)
 * 
 * Matches Minima Java implementation which uses BouncyCastle:
 * - Winternitz.java: WINTERNITZ_VALUE = 8
 * - WinternitzOTSignature.java: w=8 means 8 BITS per digit (not base-8)
 * - SHA3-256 hash function (mdsize = 32 bytes)
 * 
 * Chain count calculation (from WinternitzOTSignature constructor):
 *   messagesize = ((mdsize << 3) + w - 1) / w = (256 + 7) / 8 = 32
 *   checksumsize = getLog((messagesize << w) + 1) = getLog(8193) = 14 bits
 *   keysize = messagesize + (checksumsize + w - 1) / w = 32 + (14 + 7) / 8 = 34
 * 
 * So L = 34 chains total, each chain value is 0-255 (8-bit digit)
 */
export type ParamSet = {
  name: 'minima';
  n: 256;           // hash output bits
  w: 8;             // Winternitz parameter: 8 BITS per digit
  L: 34;            // Total chains: 32 message + 2 checksum
  messageSize: 32;  // Number of message digits
  checksumSize: 14; // Checksum bits
  checksumDigits: 2; // Number of checksum digits
  maxDigit: 255;    // Maximum digit value (2^w - 1)
};

export const WOTS_MINIMA: ParamSet = { 
  name: 'minima', 
  n: 256, 
  w: 8, 
  L: 34,
  messageSize: 32,
  checksumSize: 14,
  checksumDigits: 2,
  maxDigit: 255,
};

// Legacy aliases for backwards compatibility during migration
export const WOTS_V1_DEV = WOTS_MINIMA;
export const WOTS_V2_SPEC = WOTS_MINIMA;

export function getParamSet(_env?: string): ParamSet {
  return WOTS_MINIMA;
}

File: packages/core-wasm/src/params.rs
/// WOTS+ parameter set for Minima blockchain.
///
/// Winternitz parameter w=8 (8 bits per digit), n=256 (SHA3-256 output),
/// L=34 chains (32 message + 2 checksum).
///
/// This matches the Java BouncyCastle-compatible implementation byte-for-byte.

/// Winternitz parameter — number of bits per digit.
pub const WOTS_W: usize = 8;

/// Hash output length in bytes (SHA3-256).
pub const WOTS_N: usize = 32;

/// Number of Winternitz chains: 32 message digits + 2 checksum digits.
pub const WOTS_L: usize = 34;

/// Maximum value of a single Winternitz digit (2^w - 1).
pub const MAX_DIGIT: u16 = 255;

/// Total number of signatures per TreeKey (64 addresses × 64 L1 × 64 L2).
pub const MAX_SIGNATURES: usize = 262_144;

/// Number of levels in the TreeKey hierarchy.
pub const SIGNATURE_LEVELS: usize = 3;

/// Number of children per TreeKeyNode.
pub const NODE_CHILDREN: usize = 64;

/// Minima address prefix.
pub const ADDRESS_PREFIX: &str = "Mx";

/// Network ID for mainnet.
pub const NETWORK_ID: u8 = 1;

/// Full WOTS signature length in bytes (L × N).
pub const SIGNATURE_LENGTH: usize = WOTS_L * WOTS_N;

/// Full WOTS public key length in bytes (L × N).
pub const PUBLIC_KEY_LENGTH: usize = WOTS_L * WOTS_N;

/// WOTS public key digest length (SHA3-256 of full public key).
pub const PK_DIGEST_LENGTH: usize = 32;

/// Seed length in bytes.
pub const SEED_LENGTH: usize = 32;

/// BIP39 mnemonic word count.
pub const MNEMONIC_WORD_COUNT: usize = 24;

/// BIP39 entropy bits.
pub const MNEMONIC_ENTROPY_BITS: usize = 256;
